### PR TITLE
:bug: fix and result of multicheck

### PR DESCRIPTION
**Pull request by asraa**
Signed-off-by: Asra Ali <asraa@google.com>

* **Please check if the PR fulfills these requirements**
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] PR title follows the guidelines defined in  https://github.com/ossf/scorecard/blob/main/CONTRIBUTING.md#pr-process

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Fixes MultiCheckAnd logic. If there is a failing result with confidence 7 and another with 10, multicheck and should return the worst, as in, the 10.

* **What is the current behavior?** (You can also link to an open issue here)
MultiCheckAnd will return the least confident of failing results. 

* **What is the new behavior (if this is a feature change)?**
MultiCheckAnd will return the most confident of failing results. 

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
No

* **Other information**:

### CODE REVIEW

File: checker/check_result.go [hunk 1]
```diff
@@ -92,3 +92,27 @@ func MakeProportionalResult(name string, numerator int, denominator int,
...
+
+// MultiCheckAnd means all checks must succeed. This returns a conservative result
+// where the worst result is returned.
+func MakeAndResult(checks ...CheckResult) CheckResult {
+	minResult := CheckResult{
+		Pass:       true,
+		Confidence: MaxResultConfidence,
+	}
+
+	for _, result := range checks {
```
Comment (laurentsimon) [inline checker/check_result.go:104]: can you also patch function [MultiCheckAnd()](https://github.com/ossf/scorecard/blob/main/checker/check_runner.go#L109) to return `MakeAndResult`?

**Review by laurentsimon — APPROVED** [review]

File: checker/check_result.go [hunk 2]
```diff
@@ -92,3 +92,27 @@ func MakeProportionalResult(name string, numerator int, denominator int,
 		Confidence: MaxResultConfidence - int(actual*MaxResultConfidence),
 	}
 }
+
+// MultiCheckAnd means all checks must succeed. This returns a conservative result
```
Comment (azeemshaikh38) [inline checker/check_result.go:96]: Update description, its still for `MultiCheckAnd`

File: checker/check_result.go [hunk 3]
```diff
@@ -92,3 +92,27 @@ func MakeProportionalResult(name string, numerator int, denominator int,
...
+		Pass:       true,
+		Confidence: MaxResultConfidence,
+	}
+
+	for _, result := range checks {
+		if minResult.Name == "" {
+			minResult.Name = result.Name
+		}
+		switch result {
+		case Bool2int(result.Pass) < Bool2int(minResult.Pass):
```
Comment (azeemshaikh38) [inline checker/check_result.go:109]: Instead make a fn `isMinResult()` where all this comparison logic lies. Then update this to:

```
if isMinResult(result, minResult) {
  minResult = result
}
```

File: checker/check_result_test.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+// limitations under the License.
+
+package checker
+
+import (
+	"errors"
+	"testing"
+)
+
+const CheckTest = "Check-Test"
```
Comment (azeemshaikh38) [inline checker/check_result_test.go:22]: Change to `checkTest` and `errTest` since these are not being exported.

File: checker/check_runner.go [hunk 5]
```diff
@@ -118,8 +118,11 @@ func MultiCheckAnd(fns ...CheckFn) CheckFn {
 			if minResult.Name == "" {
 				minResult.Name = result.Name
 			}
-			if Bool2int(result.Pass) < Bool2int(minResult.Pass) ||
-				result.Confidence < MaxResultConfidence {
+			if Bool2int(result.Pass) < Bool2int(minResult.Pass) {
```
Comment (azeemshaikh38) [inline checker/check_runner.go:121]: +1 to @laurentsimon comment. Use `MakeAndResult` here, so we don't need to re-write this logic.

**Review by azeemshaikh38 — APPROVED** [review]

**Review by laurentsimon — APPROVED** [review]

**Pull request merged**